### PR TITLE
Make io_uring or thread AIO modes a startup time decision (vs compile time)

### PR DESCRIPTION
**Pull request by cmcfarlen**
merge THREAD and IOURING AIO_MODEs. Add supported_op to io_uring lib.
move more details into AIOCallbackInternal
fix io_uring cpp checks
add config to force io_uring on/off
fix test_AIO
add config to force aio to io_uring or threads

### CODE REVIEW

File: doc/admin-guide/files/records.yaml.en.rst [hunk 1]
```diff
@@ -5196,3 +5196,46 @@ Sockets
...
+.. ts:cv:: CONFIG proxy.config.aio.force_aio STRING NULL
+
+   (Only if io_uring is enabled in the build)
+   Normally, ATS will detect if io_uring can be used for async disk IO.  Using this config item, the AIO mode
+   can instead be specified.  The value can be one of:
+
+   ============ ======================================================================
+   Value        Description
+   ============ ======================================================================
+   ``NULL``     Use the default detection logic
```
Comment (bryancall) [inline doc/admin-guide/files/records.yaml.en.rst:5235]: Why have a NULL setting in the documentation?  Is someone supposed to set it to NULL to get auto detection?

File: doc/admin-guide/files/records.yaml.en.rst [hunk 2]
```diff
@@ -5196,3 +5196,46 @@ Sockets
...
+.. ts:cv:: CONFIG proxy.config.io_uring.wq_workers_bounded INT 0
+.. ts:cv:: CONFIG proxy.config.io_uring.wq_workers_unbounded INT 0
+
+   These settings configured the number of threads for the io_uring worker queue backend.  See the manpage for
+   io_uring_register_iowq_max_workers for more information.
+
+AIO
+===
+
+.. ts:cv:: CONFIG proxy.config.aio.force_aio STRING NULL
```
Comment (bryancall) [inline doc/admin-guide/files/records.yaml.en.rst:5226]: I would change "force_aio" to "mode" or something.  Seems kind of redundant to have aio twice.

File: src/traffic_server/traffic_server.cc [hunk 3]
```diff
@@ -2007,6 +2036,10 @@ main(int /* argc ATS_UNUSED */, const char **argv)
 
   REC_ReadConfigInteger(thread_max_heartbeat_mseconds, "proxy.config.thread.max_heartbeat_mseconds");
 
+#if TS_USE_LINUX_IO_URING
+  configure_io_uring();
+#endif
+
```
Comment (bryancall) [inline src/traffic_server/traffic_server.cc:2042]: Below on line 2076 it is going to print that it is "Using io_uring for AIO" if the server is running in the thread mode.

```
[May  8 14:20:48.641] traffic_server WARNING: <AIO.cc:241 (ink_aio_init)> Using thread for AIO
[May  8 14:20:48.641] traffic_server NOTE: <traffic_server.cc:2076 (main)> Using io_uring for AIO
[May  8 14:20:48.641] traffic_server NOTE: <traffic_server.cc:2080 (main)> io_uring: WQ workers - bounded = 128, unbounded = 256574
```

File: iocore/aio/AIO.cc [hunk 4]
```diff
@@ -190,35 +182,69 @@ ink_aio_init(ts::ModuleVersion v)
...
+  case AIOBackend::AIO_BACKEND_IO_URING:
+    use_io_uring = true;
+    break;
+  case AIOBackend::AIO_BACKEND_THREAD:
+    use_io_uring = false;
+    break;
+  }
+
+  if (use_io_uring) {
+    Warning("Using io_uring for AIO");
```
Comment (bryancall) [inline iocore/aio/AIO.cc:239]: Should this be a Note instead of Warning?

**Review by bryancall — CHANGES_REQUESTED** [review]

**Comment by bryancall** [comment]
With this PR we are breaking `--enable-experimental-linux-native-aio`.  Is that what we want to do? When will `--enable-experimental-linux-native-aio` option be removed?

File: doc/admin-guide/files/records.yaml.en.rst (hunk 1 above)
Comment (cmcfarlen) [reply doc/admin-guide/files/records.yaml.en.rst:5235]: I'll just default it to auto and leave this out

**Comment by cmcfarlen** [comment]
> With this PR we are breaking `--enable-experimental-linux-native-aio`. Is that what we want to do? When will `--enable-experimental-linux-native-aio` option be removed?

I can try to fix it if we want to keep it

**Review by bryancall — APPROVED** [review]

**Review by bryancall — APPROVED** [review]

**Pull request merged**